Adds option to use buffer ID as index

cuducos opened the pull request:
Hi and thanks for this cool plug-in : )

**Context**

I have a use case in mind where I like to know the buffer ID number.

This allows users to run commands for the specific buffer. For example, easily load a buffer into a split with `:sb 4` where 4 is the buffer ID number.

**Problem**

Currently _buftabline.nvim_ only shows the sequential index in the tabline…

**Approach**

… so I added a `buffer_id_index` option.

**How to test it**

If this is not clear, let me create an example:

1. Open `nvim`
2. Start a new file/buffer `:e testfile`
3. Start another file/buffer `:e testfile2`
4. Close the last buffer `:bw`
5. And start yet another one `:e testfile3`

With the current and default options, the tabline should look like:

```
1. testfile   2. testfile3
```

With `buffer_id_index = true` it should result in something like:

```
1. testfile   3. testfile3
```

**Doubts**

I am pretty new to Lua and I couldn't make the tests run, but I did my best to add a test case.

<details>

```
$ make test
nvim --headless --noplugin -c "lua require(\"plenary.test_harness\").test_directory_command('test {minimal_init = \"test/minimal.vim\"}')"
Starting...Scheduling: test/bufferline_spec.lua
Scheduling: test/buffers_spec.lua
Scheduling: test/set-hlgroup_spec.lua
Scheduling: test/options_spec.lua
Scheduling: test/auto-hide_spec.lua
Scheduling: test/buftabline_spec.lua
Scheduling: test/set-maps_spec.lua
Scheduling: test/add-commands_spec.lua
Scheduling: test/utils_spec.lua
[nvim-treesitter] [0/1] Downloading...
[nvim-treesitter] [0/1] Creating temporary directory
[nvim-treesitter] [0/1] Extracting...
[nvim-treesitter] [0/1] Compiling...
Error detected while processing command line:
Error during compilation
```

</details>

Is that useful for the plugin? I'm happy to help : )

Just let me know and I'll try to find out why test is not running.

cuducos commented:
I am really sorry.

I re-read the `README.md` and found this bit:

> Non-ordinal numbers. LSP servers and plugins constantly create and
> delete temporary buffers, so I don't personally see the point.

So I think this is not in the interest of this plug-in. Gonna keep my fork then : )
Feel free to close this PR and sorry about the noise!

jose-elias-alvarez commented:
I can see that it's a useful feature for users who prefer to interact with buffers via commands, so I'm not opposed to it, actually (I was maybe more of a purist when I wrote the readme). 

If you're still interested in having the PR merged, could you move the logic to the `get_buffers` method in `buffers.lua`? Could you also let me know what error you're seeing when you attempt to run tests? Your test is working fine on my end, though it's returning the name as `5. testfile` (I don't think I'll ever understand Vim's internal buffer numbering).

cuducos commented:
Yay, really happy to see we can have it here ✨ 

Regarding the error, it just says `Error during compilation` with no specific error message (one can check the whole output in the opening message of this PR). I'll try to find out more about the error, move the logic to `get_buffers` in `buffers.lua` and double-check the tests!

Many thanks for the feedback : )

jose-elias-alvarez commented:
Ah, I see that the test output was in your original message. It seems like it's trying to load your Neovim configuration and failing at the point where it tries to install Treesitter. Could you try modifying the Makefile and change `--noplugin` to `-u NONE` and see if that makes a difference? (If it does, we can put that into a separate commit.)

cuducos commented:
> it's trying to load your Neovim configuration and failing at the point where it tries to install Treesitter

Thanks for the heads up! I fixed something I was postponing in the Treesitter settings and it got through this point. I had to manually set `rtp` in `test/minimal.vim` (probably because I did some hacking to have two versions of NeoVim available in my machine) and now I can run tests!

<details>

```diff
-set rtp+=../plenary.nvim
+" set rtp+=../plenary.nvim
+set rtp+=/Users/cuducos/.config/nvim/plugged/plenary.nvim
+set rtp+=/Users/cuducos/projects/buftabline.nvim
```

</details>

**Anyway, moving on to what matters the most**

In 7a1f61f I suggested the following:

1. I moved the logic to `buffers.lua`
2. I changed the _get_buf_numbers should get table of ordinal buffer numbers_ test case so instead of generating 5 buffers, it generates 5 and wipes 2 of them — and we expect the indexes to be 1, 2 and 3
3. I added the _get_buf_numbers should get a table of non-ordinal buffer numbers when buffer_id_index is true_ to use the same buffer creation strategy (creating 5 buffers, wiping 2 of them) and expect to see no buffer with index 2 (so we check it's not ordinal)

jose-elias-alvarez commented:
Looks good, and I think the change to the other test is sensible, too. 

There are a few methods in `init.lua` that depend on buffer indexes, and I think we should have those adapt to the new option for consistency.

1. `next_buffer` and `prev_buffer`: the easiest solution would be to have these call `:bnext` and `:bprev` if the option is set.
2. `buftarget`: when using buffer IDs, this can simply call `vim.cmd(command .. " " .. number)`.

Finally, we can remove the line from the readme (under "Non-goals") that mentions non-ordinal numbers, since it's no longer relevant. After that, it should be all set to merge.

cuducos commented:
Great! Thanks for the guidance. Not sure the code suggested in 3679cf6 is the best way to do it in Lua (in terms of expressiveness and readability). Fell free to point me in another direction : )

Also, LMK if you would like me to squash commits in order to have a more concise Git history.

jose-elias-alvarez commented:
I think the slightly more idiomatic way to do it would be an early return, but I'm not too bothered by it for such a small project. Everything looks good on my end. Thanks for working on this!

cuducos commented:
> I think the slightly more idiomatic way to do it would be an early return

Hum… thanks for sharing : )
Maybe a refactor next time…
Thanks for allowing this new feature — very helpful here!
